### PR TITLE
Bump mypy from 1.4.1 to 1.5.1.

### PR DESCRIPTION
**Pull request by clokep**
#16291 lets us bump this, unfortunately it didn't seem to tighten anything up.

### CODE REVIEW

File: poetry.lock [hunk 1]
```diff
@@ -2077,23 +2077,15 @@ files = [
     {file = "PyYAML-6.0.1-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl", hash = "sha256:69b023b2b4daa7548bcfbd4aa3da05b3a74b772db9e23b982788168117739938"},
     {file = "PyYAML-6.0.1-cp310-cp310-manylinux_2_17_s390x.manylinux2014_s390x.whl", hash = "sha256:81e0b275a9ecc9c0c0c07b4b90ba548307583c125f54d5b6946cfee6360c733d"},
     {file = "PyYAML-6.0.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl", hash = "sha256:ba336e390cd8e4d1739f42dfe9bb83a3cc2e80f567d8805e11b46f4a943f5515"},
-    {file = "PyYAML-6.0.1-cp310-cp310-musllinux_1_1_x86_64.whl", hash = "sha256:326c013efe8048858a6d312ddd31d56e468118ad4cdeda36c719bf5bb6192290"},
```
Comment (clokep) [inline poetry.lock:2080]: I'm guessing this is yaml/pyyaml#737?

**Comment by DMRobertson** [comment]
> unfortunately it didn't seem to tighten anything up.

It seemed to pickup a couple of unused type-ignores at least!

**Comment by clokep** [comment]
> It seemed to pickup a couple of unused type-ignores at least!

Huh, why didn't I see those locally?

**Comment by clokep** [comment]
> Huh, why didn't I see those locally?

Apparently I do see them. 😕 Well, regardless they should be fixed now.

File: synapse/logging/opentracing.py [hunk 2]
```diff
@@ -991,11 +991,7 @@ def _decorator(func: Callable[P, R]) -> Callable[P, R]:
         if not opentracing:
             return func
 
-        # type-ignore: mypy seems to be confused by the ParamSpecs here.
-        # I think the problem is https://github.com/python/mypy/issues/12909
```
Comment (DMRobertson) [inline synapse/logging/opentracing.py:995]: Aha, nice to see this get fixed up.

Comment (DMRobertson) [reply synapse/logging/opentracing.py:995]: Oh, is the point that [`strict_concatenate`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-strict_concatenate) was getting in the way of this?

**Review by DMRobertson — APPROVED** [review]

File: synapse/logging/opentracing.py (hunk 2 above)
Comment (clokep) [reply synapse/logging/opentracing.py:995]: Errr, I got a warning saying it was deprecated (and replaced with `--extra-checks` -- that doesn't seem settable on the command line). But I think you're right.

**Pull request merged**